address webfinger feedbacks (#3446)
kanidm/kanidm: book/src/integrations/oauth2.md

## WebFinger

> [!NOTE]
>
> WebFinger support requires Kanidm v1.5.1 or later.

[WebFinger](https://datatracker.ietf.org/doc/html/rfc7033) provides a mechanism
for discovering information about entities at a well-known URL

> (`http://example.com/.well-known/webfinger`).

Reviewer: [Hostname] instead of example.com. else we are implying the well known URL has to be under example.com

Author: Done

It can be used by a WebFinger client to
[discover the OIDC issuer URL](https://datatracker.ietf.org/doc/html/rfc7033#section-3.1)
of an identity provider from the hostname alone, and seems to be intended to
support dynamic client registration flows for large public identity providers.

> Kanidm v1.5.1 and later can respond to WebFinger requests, using a user's SPN as

Reviewer: Agree with yaleman. No need to add the version a second time. We already point it out above

Author: Done

> the account (eg: `user@idm.example.com`). This *does not* match on email
> addresses, because they are not required by Kanidm nor guaranteed to be unique.

Reviewer: ```suggestion
the account (eg: `user@idm.example.com`). This may not match their email
address possibly leading to problems in the client. Because the email attribute is not required by Kanidm nor guaranteed to be unique, it is up to the administrator to ensure that it matches the spn if the service requires that.
```

Author: I did some more reading, I was incorrect to suggest that's because of how Kanidm works with email addresses. [`acct` is explicitly "maybe not email address"](https://datatracker.ietf.org/doc/html/rfc7565):

> It is not assumed that an entity will necessarily be able to interact with a user's account using any particular application protocol, such as email

However, I strongly suspect that some clients will assume a vaguely email address shaped account ID *is* a valid email address, so I'll update this.

Author: Now reworded in light of the above.

> When setting up an (enterprise) application to authenticate with Kanidm,

Reviewer 2: ```suggestion
When setting up an application to authenticate with Kanidm,
```

Doesn't matter what kind of application.

Author: Done

WebFinger **does not add any security** over configuring an OpenID Discovery
URL directly. In an OIDC context, the specification makes a number of flawed
assumptions which make it difficult to use with Kanidm:

* WebFinger assumes that the identity provider will give the same `iss`
  (issuer) and OpenID Discovery document, including all URLs and signing keys,
  for *all* OAuth 2.0/OIDC clients.

  Kanidm uses *different* `iss` (issuer), signing keys, and some client-specific
  endpoint URLs, which ensures that tokens can only be used with their intended

>   service. *Changing this behaviour would reduce Kanidm's security.*

Reviewer: ```suggestion
  service.
```

There is no way kani is changing this so I wouldn't even mention changing it.

Author: Done

* WebFinger endpoints must be served at the *root* of the domain of a user's
  SPN (ie: information about the user with SPN `user@idm.example.com` is at

>   `https://idm.example.com/.well-known/webfinger?resource=acct%3Auser%40idm.example.com&rel=...`).

Author: Done

  ```caddy
  # Match on a prefix, and use {uri} to preserve all query parameters.
  # This only supports *one* client.
  example.com {
    redir /.well-known/webfinger https://idm.example.com/oauth2/openid/:client_id:{uri} 307
  }
  ```

  If you have *multiple* WebFinger clients, it will need to map some other
  property of the request (such as a source IP address or `User-Agent` header)
  to a client ID, and redirect to the appropriate WebFinger URL for that client.

* Kanidm responds to *all* WebFinger queries with
  [an Identity Provider Discovery for OIDC URL](https://datatracker.ietf.org/doc/html/rfc7033#section-3.1),
  **ignoring** any supplied
  [`rel` parameter](https://datatracker.ietf.org/doc/html/rfc7033#section-4.3).

  If you want to use WebFinger in any *other* context on Kanidm's hostname,
  you'll need a load balancer in front of Kanidm which matches on some property
  of the request.

  WebFinger clients *may* omit the `rel=` parameter, so if another service has
  relations for an `acct:` entity and a client *does not* supply the `rel=`
  parameter, your load balancer will need to merge JSON responses from Kanidm
  and the other service(s).

> 

Author: Depends on the client, really.

In the Tailscale case, it seems like they only check WebFinger during setup, then assume that a once-valid WebFinger claim for an account ID is a valid claim for the whole domain for all time. So you could absolutely host a static file for a few minutes, then delete it.

If a client checks WebFinger on each login attempt (eg: it has a login form which prompts for an email-like username, then attempts to pick an authentication methods based on a WebFinger response), then it'll need to stay up, and provide a correct account ID in the response.

And yes, this sucks!

> 

Reviewer: Isn't mean temporarily hosting a static file. I was referring to a permanent static file containing the webfinger URLs for all kinds of rels including profile picture, blog & contact.

Its pretty unconventional so its probably best not to include it.

> 

Author: Added in a suggestion to serve statically.

> Because of these issues, we recommend that (enterprise) applications support

Reviewer 2: ```suggestion
Because of these issues, we recommend that applications support
```

Author: Done

entities at a well-known URL (`https://{hostname}/.well-known/webfinger`).

It can be used by a WebFinger client to
[discover the OIDC issuer URL][webfinger-oidc] of an identity provider from the
hostname alone, and seems to be intended to support dynamic client registration
flows for large public identity providers.

Kanidm v1.5.1 and later can respond to WebFinger requests, using a user's SPN as
part of [an `acct` URI][rfc7565] (eg: `acct:user@idm.example.com`). While SPNs
and `acct` URIs look like email addresses, [as per RFC 7565][rfc7565s4], there
is no guarantee that it is valid for any particular application protocol, unless
an administrator explicitly provides for it.

When setting up an application to authenticate with Kanidm, WebFinger **does not
add any security** over configuring an OpenID Discovery URL directly. In an OIDC
context, the specification makes a number of flawed assumptions which make it
difficult to use with Kanidm:

* WebFinger assumes that the identity provider will give the same `iss`
  (issuer) and OpenID Discovery document, including all URLs and signing keys,
  for *all* OAuth 2.0/OIDC clients.

>   Kanidm uses *different* `iss` (issuer), signing keys, and some client-specific

Reviewer 3: Different client signing keys may be better.

Author: I reworded this to "uses *client-specific* ... token signing keys"

  client ID in the request, so there is no way to tell them apart.

  As a result, Kanidm *does not* provide a WebFinger endpoint at its root URL,
  because it could report an incorrect `iss` (issuer) and lead the client to an
  incorrect OIDC discovery document.

  You will need a load balancer in front of Kanidm's HTTPS server to send a HTTP
  307 redirect to the appropriate
  `/oauth2/openid/:client_id:/.well-known/webfinger` URL, *while preserving all
  query parameters*. For example, with Caddy:

  ```caddy
  # Match on a prefix, and use {uri} to preserve all query parameters.
  # This only supports *one* client.
  example.com {
    redir /.well-known/webfinger https://idm.example.com/oauth2/openid/:client_id:{uri} 307
  }
  ```

  If you have *multiple* WebFinger clients, it will need to map some other
  property of the request (such as a source IP address or `User-Agent` header)
  to a client ID, and redirect to the appropriate WebFinger URL for that client.

* Kanidm responds to *all* WebFinger queries with
  [an Identity Provider Discovery for OIDC URL][webfinger-oidc], **ignoring**
  any supplied [`rel` parameter][webfinger-rel].

>   If you want to use WebFinger in any *other* context on Kanidm's hostname,

Reviewer: ```suggestion
  [`rel` parameters](https://datatracker.ietf.org/doc/html/rfc7033#section-4.3).
```

Author: I don't think your suggestion is showing up GitHub as you intended:

<img width="741" alt="Screenshot 2025-02-21 at 14 02 17" src="https://github.com/user-attachments/assets/f9362002-6dfe-499c-8673-d3871e9298d3" />

I think you actually want me to replace "any supplied `rel` parameter" with "`rel` parameters" (less words, also the spec allows it to be specified multiple times).

Author: I've added the above change.

## WebFinger

[WebFinger](https://datatracker.ietf.org/doc/html/rfc7033) provides a mechanism
for discovering information about entities at a well-known URL
(`https://{hostname}/.well-known/webfinger`).

It can be used by a WebFinger client to
[discover the OIDC issuer URL](https://datatracker.ietf.org/doc/html/rfc7033#section-3.1)
of an identity provider from the hostname alone, and seems to be intended to
support dynamic client registration flows for large public identity providers.

> Kanidm v1.5.1 and later can respond to WebFinger requests, using a user's SPN as

Reviewer: I preferred the version in a little note at the top so admins don't need to read the entire text to know what version they need to update to

Author: @yaleman any opinions on this? 😅

Reviewer 3: It's 1.5.0, but also I think we don't need it because docs are published per-server-version, so when we merge this it won't show until 1.6 iirc.
